In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [6]:
# normalize the data
scaled_market_data = StandardScaler().fit_transform(df_market_data[['price_change_percentage_24h', 'price_change_percentage_7d', 'price_change_percentage_14d', 'price_change_percentage_30d', 'price_change_percentage_60d', 'price_change_percentage_200d', 'price_change_percentage_1y']])

In [7]:
# Create a DataFrame with the scaled data
scaled_df_market_data = pd.DataFrame(scaled_market_data, 
                                         columns=['price_change_percentage_24h', 'price_change_percentage_7d', 
                                         'price_change_percentage_14d', 'price_change_percentage_30d', 
                                         'price_change_percentage_60d', 'price_change_percentage_200d', 
                                         'price_change_percentage_1y'])


# Copy the crypto names from the original DataFrame
scaled_df_market_data['coin_id'] = df_market_data.index


# Set the coin_id column as index
scaled_df_market_data = scaled_df_market_data.set_index('coin_id')


# Display the scaled DataFrame
scaled_df_market_data.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [8]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [9]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=42)
    k_model.fit(scaled_df_market_data)
    inertia.append(k_model.inertia_)

In [10]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {'k': k, 'inertia': inertia}


# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow_data)

In [12]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_df.hvplot.line(
    x = 'k',
    y = 'inertia',
    title = 'Elbow Curve',
    xticks = k
)


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** While the visual analysis indicates that k=4 appears to be the optimal cluster count, standard methodology recommends validating this by testing neighboring values (e.g., k=3 and k=5) to ensure the chosen value truly maximizes performance metrics and isn’t a local anomaly. This practice helps confirm robustness and reduces the risk of overfitting or underfitting the model.

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [13]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=42)

In [14]:
# Fit the K-Means model using the scaled DataFrame
model.fit(scaled_df_market_data)

KMeans(n_clusters=4, random_state=42)

In [15]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
kmeans_prediction = model.predict(scaled_df_market_data)

# Print the resulting array of cluster values.
print(kmeans_prediction)

[0 0 2 2 0 0 0 0 0 2 2 2 2 0 2 0 2 2 0 2 2 0 2 2 2 2 2 2 0 2 2 2 3 0 2 2 1
 2 2 2 2]


In [16]:
# Create a copy of the scaled DataFrame
scaled_df_copy = scaled_df_market_data.copy()

In [17]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
scaled_df_copy['Predictions'] = kmeans_prediction

# Display the copy of the scaled DataFrame
scaled_df_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Predictions  
coin_id                                                
bitcoin                        -0.251637            0  
ethereum                       -0.199352            0  
tether                         -0.282061            2  
ripple                         -0.295546            2  
bitcoin-cash                   -0.270317            0

In [18]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
scaled_plot = scaled_df_copy.hvplot.scatter(
    x = 'price_change_percentage_24h',
    y = 'price_change_percentage_7d',
    by = 'Predictions'
)

scaled_plot

:NdOverlay   [Predictions]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

---

### Optimize Clusters with Principal Component Analysis.

In [19]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [20]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
crypto_data = pca.fit_transform(scaled_df_copy)

# View the scaled PCA data
crypto_data[:5]

array([[ 1.58513942,  0.13439444,  0.545621  ],
       [ 1.32258066,  0.05449393,  1.11570854],
       [-0.26633536, -0.53640937, -0.79694588],
       [-0.25722513, -0.59201286, -0.6465287 ],
       [ 2.95224985,  0.42237155,  1.26789353]])

In [21]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
pca.explained_variance_

array([3.00217653, 2.62223875, 1.55078065])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:**   3.00217653+2.62223875+1.55078065= (total) 7.17519593
​	
 

In [22]:
scaled_df_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Predictions  
coin_id                                                
bitcoin                        -0.251637            0  
ethereum                       -0.199352            0  
tether                         -0.282061            2  
ripple                         -0.295546            2  
bitcoin-cash                   -0.270317            0

In [24]:
# Create a new DataFrame with the PCA data.
crypto_pca_df = pd.DataFrame(crypto_data,
                             columns=['PCA1', 'PCA2', 'PCA3'])

# Copy the crypto names from the original scaled DataFrame
crypto_pca_df['coin_id'] = scaled_df_copy.index

# Set the coin_id column as index
crypto_pca_df = crypto_pca_df.set_index('coin_id')

# # Display the scaled PCA DataFrame
crypto_pca_df.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin       1.585139  0.134394  0.545621
ethereum      1.322581  0.054494  1.115709
tether       -0.266335 -0.536409 -0.796946
ripple       -0.257225 -0.592013 -0.646529
bitcoin-cash  2.952250  0.422372  1.267894

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [25]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1,11))

In [27]:
# Create an empty list to store the inertia values
inertia_list = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k_values:
    k_model = KMeans(n_clusters=i, random_state=42)
    k_model.fit(crypto_pca_df)
    inertia_list.append(k_model.inertia_)

In [28]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_2 = {'k': k_values, 'inertia': inertia_list}
# Create a DataFrame with the data to plot the Elbow curve
elbow_df_2 = pd.DataFrame(elbow_data_2)


In [29]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_df_2.hvplot.line(
    x= 'k',
    y= 'inertia',
    title = 'Elbow Curve 2',
    xticks = k_values
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for K is 4.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:**  Does not seem to differ from the best k value using the original data.

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [30]:
# Initialize the K-Means model using the best value for k
model_2 = KMeans(n_clusters=4, random_state=42)

In [31]:
# Fit the K-Means model using the PCA data
model_2.fit(crypto_pca_df)

KMeans(n_clusters=4, random_state=42)

In [33]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
kmeans_prediction_2 = model_2.predict(crypto_pca_df)

# Print the resulting array of cluster values.
print(kmeans_prediction_2)

[0 0 1 1 0 0 0 0 0 1 1 1 1 0 1 0 1 1 0 1 1 0 1 1 1 1 1 1 0 1 1 1 3 0 1 1 2
 1 1 1 1]


In [35]:
# Create a copy of the scaled PCA DataFrame
copy_crypto_pca_df = crypto_pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters

copy_crypto_pca_df['Prediction'] = kmeans_prediction_2
# Display the copy of the scaled PCA DataFrame
copy_crypto_pca_df.head()

PCA1      PCA2      PCA3  Prediction
coin_id                                               
bitcoin       1.585139  0.134394  0.545621           0
ethereum      1.322581  0.054494  1.115709           0
tether       -0.266335 -0.536409 -0.796946           1
ripple       -0.257225 -0.592013 -0.646529           1
bitcoin-cash  2.952250  0.422372  1.267894           0

In [36]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
pca_plot = copy_crypto_pca_df.hvplot.scatter(
    x='PCA1',
    y='PCA2',
    by = 'Prediction',
    hover_cols = ['index'],
    title='PCA Scatter Plot of Crypto Clusters'
)

pca_plot

:NdOverlay   [Prediction]
   :Scatter   [PCA1]   (PCA2)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [37]:
# Composite plot to contrast the Elbow curves
scaled_plot

:NdOverlay   [Predictions]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

In [38]:
# Composite plot to contrast the clusters
pca_plot

:NdOverlay   [Prediction]
   :Scatter   [PCA1]   (PCA2)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:**  After reducing the number of clusters, the data now shows much clearer groupings. Unlike the pre-PCA version, which had difficulty revealing distinct clusters, the post-PCA version displays well-defined clusters and highlights outliers more effectively.